cpu/native/ng_tapnet/ng_tapnet.c

```diff
@@ -0,0 +1,615 @@
+     * available */
+
+    fd_set rfds;
+
+    while (1) {
+        FD_ZERO(&rfds);
+        FD_SET(_native_tap_fd, &rfds);
+
+        if (real_select(_native_tap_fd + 1, &rfds, NULL, NULL, NULL) == 1) {
+            kill(parent, SIGIO);
+        }
+        else {
+            kill(parent, SIGKILL);
+            err(EXIT_FAILURE, "osx_sigio_child: select");
+        }
+
+        pause();
+    }
+}
+#endif
+
+static int _marshall_ethernet(ng_tapnet_t *dev, uint8_t *buffer, ng_pktsnip_t *pkt)
+{
+    int data_len = 0;
+    struct ether_header *hdr = (struct ether_header *)buffer;
+    ng_netif_hdr_t *netif_hdr;
+    ng_pktsnip_t *ptr = pkt, *payload;
+
+    if (ptr == NULL) {
+        DEBUG("ng_tapnet: pkt was NULL");
+        return -EINVAL;
+    }
+
+    while (ptr->next != NULL) {
+        payload = ptr;
+        ptr = ptr->next;
+    }
+
+    if (ptr->size != (sizeof(ng_netif_hdr_t) + ETHER_ADDR_LEN)) {
+        DEBUG("ng_tapnet: First header was not generic netif header\n");
+        return -EBADMSG;
+    }
+
+    switch (payload->type) {
+#ifdef MODULE_NG_IPV6
+
+        case NG_NETTYPE_IPV6:
+            DEBUG("ng_tapnet: use ether type IPV6\n");
+            hdr->ether_type = ETHERTYPE_IPV6;
+            break;
+#endif
+
+        default:
+            DEBUG("ng_tapnet: use ether type unknown, use 0xffff\n");
+            hdr->ether_type = 0xffff;
+            break;
+    }
+
+    netif_hdr = ptr->data;
+
+    memcpy(hdr->ether_shost, dev->addr, ETHER_ADDR_LEN);
+    memcpy(hdr->ether_dhost, ng_netif_hdr_get_dst_addr(netif_hdr),
+           netif_hdr->dst_l2addr_len);
+    DEBUG("ng_tapnet: send to %02x:%02x:%02x:%02x:%02x:%02x\n",
+          hdr->ether_dhost[0], hdr->ether_dhost[1], hdr->ether_dhost[2],
+          hdr->ether_dhost[3], hdr->ether_dhost[4], hdr->ether_dhost[5]);
+
+    LL_DELETE(pkt, ptr);
+    ng_pktbuf_release(ptr);
+
+    buffer += ETHER_HDR_LEN;
+    data_len += ETHER_HDR_LEN;
+
+    while (pkt != NULL) {
+        LL_SEARCH_SCALAR(pkt, ptr, next, NULL);
+
+        if ((data_len + ptr->size) > ETHER_MAX_LEN) {
+            DEBUG("ng_tapnet: Packet too big for ethernet frame\n");
+            return -ENOBUFS;
+        }
+
+        memcpy(buffer, pkt->data, pkt->size);
+        buffer += ptr->size;
+        data_len += ptr->size;
+
+        LL_DELETE(pkt, ptr);
+        ng_pktbuf_release(ptr);
+    }
+
+    /* Pad to minimum payload size.
+     * Linux does this on its own, but it doesn't hurt to do it here.
+     * As of now only tuntaposx needs this. */
+    if (data_len < ETHERMIN) {
+        DEBUG("padding data!(%d -> ", data_len);
+        memset(buffer, 0, ETHERMIN - data_len);
+        data_len = ETHERMIN;
+        DEBUG("%d)\n", data_len);
+    }
+
+    return data_len;
+}
+
+void _trigger_isr_event(void)
+{
+    msg_t msg;
+
+    DEBUG("ng_tapnet: Trigger ISR event\n");
+
+    /* TODO: check whether this is an input or an output event
+       TODO: refactor this into general io-signal multiplexer */
```

Comment:
@LudwigOrtmann I pretty much copied this from `nativenet`'s `_native_handle_tap_input()`. Is there a way to get the signal triggering file descriptor so I can find the device that triggered this?